### PR TITLE
feat(hip): Add Implementation of hipLibraryGetModule API

### PR DESCRIPTION
**Pull request by satyanveshd**
## Motivation

Adds the implementation for HIP API hipLibraryGetModule for the corresponding CUDA API cuLibraryGetModule.

## Technical Details

Adds relevant helper function Module() to fetch the corresponding module from the dynamic code object. Also adds RegisterLibraryModule, UnregisterLibraryModule to register and unregister the modules from the dynamic code object map.

## Issue Tracking

JIRA ID: AIRUNTIME-2611

## Test Plan

Add few hip catch tests verifying different positive and negative usages of this API, like fetching the module and using that to get the functions, kernels etc.

## Test Result

PASS

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

**Comment by danielbolsson** [comment]
Nice — the HRR side is half in place already: `hipLibraryGetModule` in
`NOOP_PLAYBACK_APIS` is the right classification (same reason as its
`hipLibrary*` siblings — `hipLibrary_t` handles aren't tracked at playback),
and appending `hipLibraryGetModule_fn` at the end of `HipDispatchTable` keeps
existing HRR API IDs stable, so older archives stay readable.

What's missing is the generated half. HRR's generated sources are committed
artifacts, not build-time outputs — nothing in CMake runs the generator — so
the `gen_hrr_api_args.py` edit has no effect on its own. Could you run it and
include the output in this PR?

      python3 projects/hrr/tools/gen_hrr_api_args.py

It updates three tracked files:

  - `projects/hrr/include/hrr/hrr_api_args.h` — `hrr_args_hipLibraryGetModule`,
    `HRR_API_HIPLIBRARYGETMODULE = 553`, `HRR_API_COUNT` 553 → 554
  - `projects/clr/hipamd/src/hrr/hip_capture_generated.cpp` — capture shim +
    table entry
  - `projects/hrr/playback/hip_playback_generated.cpp` — no-op playback handler,
    dispatch entry, min-payload entry

Until those land the capture runtime has no shim for the new API and calls are
silently not recorded — no build error, they just never show up in a capture.

Two notes:

  - Regenerating also picks up one unrelated line already stale on develop
    (`hipDeviceProp_t` → `hipDeviceProp_tR0600` in the comment above
    `hipChooseDevice`). Including it or restricting the commit to the
    `hipLibraryGetModule` hunks are both fine.
  - Worth adding the call to `Unit_HRR_ModuleExtra_Direct` in
    `projects/hrr/tests/integration/hrr_workload_test.cc`, where the rest of the
    `hipLibrary*` family is exercised. Capture shims only record on `hipSuccess`,
    so a nullptr-handle call builds and runs but captures nothing — reusing the
    real library handle from your new `library_get_module.cc` test would actually
    cover capture and replay.

File: projects/hrr/tools/gen_hrr_api_args.py [hunk 1]
```diff
@@ -543,6 +543,7 @@
     "hipLibraryUnload",
     "hipLibraryGetKernel",
     "hipLibraryGetKernelCount",
+    "hipLibraryGetModule",
```
[the remark this answers is not in the document]
Comment (satyanveshd) [reply projects/hrr/tools/gen_hrr_api_args.py:548]: Done.

**Comment by satyanveshd** [comment]
> Nice — the HRR side is half in place already: `hipLibraryGetModule` in `NOOP_PLAYBACK_APIS` is the right classification (same reason as its `hipLibrary*` siblings — `hipLibrary_t` handles aren't tracked at playback), and appending `hipLibraryGetModule_fn` at the end of `HipDispatchTable` keeps existing HRR API IDs stable, so older archives stay readable.
> 
> What's missing is the generated half. HRR's generated sources are committed artifacts, not build-time outputs — nothing in CMake runs the generator — so the `gen_hrr_api_args.py` edit has no effect on its own. Could you run it and include the output in this PR?
> 
> ```
>   python3 projects/hrr/tools/gen_hrr_api_args.py
> ```
> 
> It updates three tracked files:
> 
> * `projects/hrr/include/hrr/hrr_api_args.h` — `hrr_args_hipLibraryGetModule`,
>   `HRR_API_HIPLIBRARYGETMODULE = 553`, `HRR_API_COUNT` 553 → 554
> * `projects/clr/hipamd/src/hrr/hip_capture_generated.cpp` — capture shim +
>   table entry
> * `projects/hrr/playback/hip_playback_generated.cpp` — no-op playback handler,
>   dispatch entry, min-payload entry
> 
> Until those land the capture runtime has no shim for the new API and calls are silently not recorded — no build error, they just never show up in a capture.
> 
> Two notes:
> 
> * Regenerating also picks up one unrelated line already stale on develop
>   (`hipDeviceProp_t` → `hipDeviceProp_tR0600` in the comment above
>   `hipChooseDevice`). Including it or restricting the commit to the
>   `hipLibraryGetModule` hunks are both fine.
> * Worth adding the call to `Unit_HRR_ModuleExtra_Direct` in
>   `projects/hrr/tests/integration/hrr_workload_test.cc`, where the rest of the
>   `hipLibrary*` family is exercised. Capture shims only record on `hipSuccess`,
>   so a nullptr-handle call builds and runs but captures nothing — reusing the
>   real library handle from your new `library_get_module.cc` test would actually
>   cover capture and replay.

@danielbolsson I have made the necessary changes and updated the hrr files as you mentioned. Please check if it looks fine. Thanks.

**Review by mangupta — COMMENTED** [review]
🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Regenerated HRR API-id enum renumbers 9 existing ids, breaking previously captured archives

The regenerated `hrr_api_id` enum inserts `HRR_API_HIPLIBRARYGETMODULE = 544` in the middle of the enumeration rather than appending it, shifting every subsequent id by one: `HRR_API_HIPPOPCALLCONFIGURATION` 544→545 through `HRR_API_HIPUNREGISTERFATBINARY` 552→553, with `HRR_API_COUNT` 553→554. The numeric id is what `write_event_raw()` stamps into each captured event, and both `hrr_api_min_payload_size[]` and `hrr_playback_dispatch[]` are indexed by it.

**Suggested fix:** Make the generator append new APIs after the existing `__hip*` compiler entries (so `HRR_API_HIPLIBRARYGETMODULE = 553`, `HRR_API_COUNT = 554`) and regenerate, preserving the ids of the nine shifted entries. If mid-enum insertion is intentional, bump the HRR archive format version and state the incompatibility so old archives are rejected explicitly rather than silently mis-decoded.



**[IMPORTANT]** Regenerated contract-coverage snapshot leaves two declared APIs neither covered nor allowlisted

The updated snapshot reports `declared_apis: 500`, `covered_apis: 487`, `uncovered_allowlisted: 11`. 500 − 487 − 11 = 2 declared APIs that are neither exercised by a contract test nor listed as an intentional exception, where the previous snapshot balanced exactly (497 − 486 − 11 = 0).

**Suggested fix:** Re-run the coverage generator against the PR head and reconcile the delta: either add contract tests for the two additional declared APIs the regeneration picked up, add justified entries to `projects/hip-tests/catch/contract/uncovered_apis.txt` (which would move `uncovered_allowlisted` 11 → 13), or correct the snapshot if the declared-API parse over-counted.



**[IMPORTANT]** New negative unit tests leave the sticky HIP error set for subsequent test cases

`Unit_hipLibraryGetModule_Negative_Parameters`, `Unit_hipLibraryGetModule_Negative_ModuleUnloadRefused`, and `Unit_hipLibraryGetModule_Negative_StaleAfterLibraryUnload` each deliberately trigger a HIP error (`hipLibraryGetModule(nullptr, lib)`, `hipLibraryGetModule(&mod, nullptr)`, `hipModuleUnload(mod)`, `hipModuleGetFunction` on a stale handle) but never consume the resulting thread-local sticky error before returning.

**Suggested fix:** Add `(void)hipGetLastError();` after each intentionally failing call in the three negative cases, matching the convention already used by the new contract tests in this PR.
